core/src/main/java/feast/core/config/SecurityConfig.java

```diff
@@ -0,0 +1,144 @@
+  AuthenticationManager authenticationManager() {
+    final List<AuthenticationProvider> providers = new ArrayList<>();
+
+    if (securityProperties.getAuthentication().isEnabled()) {
+      switch (securityProperties.getAuthentication().getProvider()) {
+        case "GoogleOpenID":
+          providers.add(
+              new GoogleOpenIDAuthenticationProvider(
+                  securityProperties.getAuthentication().getOptions()));
+          break;
+        default:
+          throw new IllegalArgumentException(
+              "Please configure an Authentication Provider if you have enabled authentication.");
+      }
+    }
+    return new ProviderManager(providers);
+  }
+
+  /**
+   * Creates an AuthenticationReader that the AuthenticationManager will use to authenticate
+   * requests
+   *
+   * @return GrpcAuthenticationReader
+   */
+  @Bean
+  @ConditionalOnProperty(prefix = "feast.security.authentication", name = "enabled")
+  GrpcAuthenticationReader authenticationReader() {
+    return new BearerAuthenticationReader(BearerTokenAuthenticationToken::new);
+  }
+
+  /**
+   * Creates a SecurityMetadataSource when authentication is enabled. This allows for the
+   * configuration of endpoint level security rules.
+   *
+   * @return GrpcSecurityMetadataSource
+   */
+  @Bean
+  @ConditionalOnProperty(prefix = "feast.security.authentication", name = "enabled")
+  GrpcSecurityMetadataSource grpcSecurityMetadataSource() {
+    final ManualGrpcSecurityMetadataSource source = new ManualGrpcSecurityMetadataSource();
+
+    // Authentication is enabled for all gRPC endpoints
+    source.setDefault(AccessPredicate.authenticated());
+
+    // The following endpoints allow unauthenticated access
+    source.set(CoreServiceGrpc.getGetFeastCoreVersionMethod(), AccessPredicate.permitAll());
+
+    return source;
+  }
+
+  /**
+   * Creates an AccessDecisionManager if authorization is enabled. This object determines the policy
+   * used to make authorization decisions.
+   *
+   * @return AccessDecisionManager
+   */
+  @Bean
+  @ConditionalOnProperty(prefix = "feast.security.authorization", name = "enabled")
+  AccessDecisionManager accessDecisionManager() {
+    final List<AccessDecisionVoter<?>> voters = new ArrayList<>();
+    voters.add(new AccessPredicateVoter());
+    return new UnanimousBased(voters);
+  }
+
+  /**
+   * Creates an AuthorizationProvider based on Feast configuration. This provider is available
+   * through the security service.
+   *
+   * @return AuthorizationProvider used to validate access to Feast resources.
+   */
+  @Bean
+  @ConditionalOnProperty(prefix = "feast.security.authorization", name = "enabled")
+  AuthorizationProvider authorizationProvider() {
+    if (securityProperties.getAuthentication().isEnabled()
+        && securityProperties.getAuthorization().isEnabled()) {
+      switch (securityProperties.getAuthorization().getProvider()) {
+        case "KetoAuthorization":
```

Comment:
same as authentication provider factory comment.

Reply:
How would you use Spring in this case?

Reply:
It's been awhile since I have done this in Spring Config but basically you dynamically create the @Bean which then can accept normal spring config:
```
@Configuration
public class AuthConfiguration
{

     @Value("${feast.auth.authorizationProviderClass}")
     private String authorizationProviderClassName;

    @Bean
     public AuthorizationProvider authorizationProvider()
     {
        Class<?> authClass = Class.forName(authorizationProviderClassName);
        AuthorizationProvider authProvider = AuthorizationProvider.class.cast(authClass.newInstance());
        return authProvider;
     }
```

Reply:
Yea that looks great, as long as failure is graceful and easy to debug.